### PR TITLE
Fixed 2 step session loading from cookie

### PR DESCRIPTION
**Pull request by KrauseFx**
3 hours of debugging and 10 new lines of code later, finally fixed the auth request for stored sessions. More information about the workaround is in the source file :+1:

### CODE REVIEW

**Comment by KrauseFx** [comment]
I should also add a test for this :+1:

File: spaceship/lib/spaceship/client.rb [hunk 1]
```diff
@@ -251,12 +251,32 @@ def send_shared_login_request(user, password)
...
+        # By default the cookie jar would generate the following header
+        #   DES5c148...=HSARM.......xaA/O69Ws/CHfQ==SRVT
+        # However we need the following
+        #   DES5c148...="HSARM.......xaA/O69Ws/CHfQ==SRVT"
+        # There is no way to get the cookie jar value with " around the value
+        # so we manually modify the cookie (only this one) to be properly escaped
+        # Afterwards we pass this value manually as a header
+        # It's not enough to just modify @cookie, it needs to be done after self.cookie
+        # as a string operation
+        important_cookie = @cookie.store.entries.find { |a| a.name.include?("DES") }
```
Comment (lacostej) [inline spaceship/lib/spaceship/client.rb]: this assumes there will always be a single cookie whose name matches this. Maybe use `select` and fail fast if we find more than one ?

Also maybe use `start_with?`

**Comment by lacostej** [comment]
I tracked this a little bit further:
* For reference, here's the HTTP Cookie specs: https://tools.ietf.org/html/rfc6265#section-4.1.1
* The culprit for stripping quotes is http-cookie: https://github.com/sparklemotion/http-cookie/issues/11
* There are a few forks: https://github.com/sparklemotion/http-cookie/network with interesting patches, the main one being https://github.com/uksa/http-cookie/commit/e54c52badce150a4b7dd416f926a8b91be9ae517 (incomplete)

The current fix is a workaround. It has the following issues:
* it only works for adding back quotes to the DES... cookie. If a new cookie has the same problem, we will have it as well
* we might have other cookie issues in the future. See https://github.com/sparklemotion/http-cookie/issues/9

It might be better to fix `http_cookie` in the long run.

**Comment by KrauseFx** [comment]
@lacostej Thanks for looking into this, great finds! I agree with you, it's a workaround, and also it requires the cookie name to start with `DES` which we don't know if it's always true. 
I'll do a fork to the fastlane organisation and point to this one instead. What do you think?

**Comment by lacostej** [comment]
@KrauseFx  I submitted a tentative fix for the quoting issue in http_cookie. See https://github.com/sparklemotion/http-cookie/pull/15

You may want to try it out ? I don't have 2 step session in my accounts so can't really easily try it. I tested it with pseudo code:

````
value = '"HSARMTKNSRVTWFla1+yO4gVPowH17VaaaxPFnUdMUegQZxqy1Ie1c2v6bM1vSOzIbuOmrl/FNenlScsd/NbF7/Lw4cpnL15jsyg0TOJwP32tC/NguPiyOaaaU+jrj4tf4uKdIywVaaaFSRVT"'

uri = URI.parse 'http://example'
cookie_str = "DES5c148586daa451e55afb017aa62418f91=#{value};Expires=Sun, 06 Nov 2020 00:28:06 GMT;Path=/"

puts cookie_str
c=nil
HTTP::Cookie.parse(cookie_str, uri) { |cookie|
  c = cookie
}

puts c.value
puts c.set_cookie_value
````

Now the `to_yaml` is still broken, mostly because I don't know how to make it work with ruby < 1.9

But you should be able t work-around it by serializing the cookie string instead. Not sure if that helps you.

**Comment by KrauseFx** [comment]
Great @lacostej, I'll give it a try on Monday! :+1:

**Comment by lacostej** [comment]
@KrauseFx I also fixed the YAML export/import i the PR.

**Comment by KrauseFx** [comment]
I'll merge this for now, and we'll see what if they approve your PR. I didn't squash since the 2 commits are separate from each other :+1: 

Thanks again @lacostej :+1:

**Pull request merged**